### PR TITLE
Split geographic coordinate strings

### PR DESCRIPTION
**Pull request by markbreedlove**
Split strings given in edm:Place (`DPLA::MAP::Place`) into latitude and longitude.

### CODE REVIEW

**Comment by anarchivist** [comment]
I'm :+1:, other than the :squash:. I'd like @AudreyAltman and/or @no-reply take a gander.

File: spec/lib/krikri/enrichments/split_coordinates_spec.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      let(:lat) { '40.0' }
+      include_examples 'nonsensical coordinates'
+    end
+
+    context 'with a latitude that splits into too many values' do
+      let(:lat) { '1.0,2.0,3.0' }
+      include_examples 'nonsensical coordinates'
+    end
+
+    context 'with a latitude containing non-decimal characters' do
```
Comment (no-reply) [inline spec/lib/krikri/enrichments/split_coordinates_spec.rb]: I suspect we'll see some data that doesn't include decimal characters, like `51N, 108W`, but I'm happy to punt this until we hit a provider where this applies.

File: lib/krikri/enrichments/split_coordinates.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    def assign_latlong!(place, lat, long)
+      # We have to assign these one at a time because we can't do this:
+      # place.lat, place.long = [[nil], [nil]]
+      # ... which results in:
+      #     "value must be an RDF URI, Node, Literal, or a valid datatype.
+      #     See RDF::Literal. You provided nil"
+      # ... although this works fine:
+      # place.lat, place.long = [['10.0'], ['11.1']]
+      place.lat = lat
```
Comment (no-reply) [inline lib/krikri/enrichments/split_coordinates.rb]: I haven't verified, but I think `place.lat, place.long = ['10.0', '11.1']` and `place.lat, place.long = [nil, nil]` will both work fine.  This already seems to be assuming that you're only working with the value of `ActiveTriples::Term#first`.

File: spec/lib/krikri/enrichments/split_coordinates_spec.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+require 'spec_helper'
+
+describe Krikri::Enrichments::SplitCoordinates do
+  it_behaves_like 'a field enrichment'
+
+  context 'with place object' do
+    let(:lat) { nil }
+    let(:long) { nil }
+    let(:place) do
```
Comment (no-reply) [inline spec/lib/krikri/enrichments/split_coordinates_spec.rb]: Minor style... maybe collapse this onto one line?

`let(:place) { build(:place, lat: lat, long: long }`

File: lib/krikri/enrichments/split_coordinates.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    #
+    # @param place [DPLA::MAP::Place]
+    # @return [DPLA::MAP::Place]
+    #
+    def enrich_value(place)
+      return place if !place.is_a? DPLA::MAP::Place
+
+      # place.lat and place.long are arrays of one value if they are assigned.
+
+      return place if place.lat.first && place.long.first  # no split required
```
Comment (no-reply) [inline lib/krikri/enrichments/split_coordinates.rb]: This requires that the mapping be to one of `lat` or `long`, but not both.  That's not what we're doing to date.  Would it be better to check whether either value contains a comma?

**Comment by AudreyAltman** [comment]
I have nothing to add to what Tom wrote.  Looking good, @mb.

**Comment by no-reply** [comment]
This looks great & is ready to go once we clarify the mapping pattern referred to here: https://github.com/dpla/KriKri/pull/144#discussion_r28072831

My notion is that looking for the delimiter is better since it makes fewer assumptions about the mapping/input data.

File: spec/lib/krikri/enrichments/split_coordinates_spec.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      end
+      it 'assigns longitude' do
+        expect(subject.enrich_value(place).long).to eq expected_long
+      end      
+    end
+
+    shared_examples 'nonsensical coordinates' do
+      it 'returns empty coordinates' do
+        rv = subject.enrich_value(place)
+        expect(rv.lat).to eq []
```
Comment (no-reply) [inline spec/lib/krikri/enrichments/split_coordinates_spec.rb]: I'm now thinking that this should retain the original values. I don't know if we want to delete coordinates that we can't recognize, but if/when we do, that seems like another enrichment. 

Thoughts @anarchivist?

**Comment by no-reply** [comment]
My changes allow mapping to both fields, giving preference to the `lat, long` format if both are given.  They also avoid destroying data unless we have successfully split the string.

**Comment by no-reply** [comment]
@AudreyAltman and @anarchivist just a quick ping to remind that there are a few commits added to this PR yesterday. If one of you can review the latest changes, i think we're ready to :ship:

**Comment by anarchivist** [comment]
:ok_hand:

**Comment by no-reply** [comment]
Squashed.

Squash

**Comment by anarchivist** [comment]
[](http://media.giphy.com/media/74eZarMmBLk8E/giphy.gif)

**Pull request merged**